test/install/test_harness.sh

```diff
@@ -0,0 +1,163 @@
+# disable using the install.sh entrypoint such that we can unit test
+# script functions without invoking main()
+TEST_INSTALL_SH=true
+
+. ../../install.sh
+set -u
+
+assertTrue() {
+  if eval "$1"; then
+    echo "assertTrue failed: $2"
+    exit 2
+  fi
+}
+
+assertFalse() {
+  if eval "$1"; then
+    echo "assertFalse failed: $2"
+    exit 2
+  fi
+}
+
+assertEquals() {
+  want=$1
+  got=$2
+  msg=$3
+  if [ "$want" != "$got" ]; then
+    echo "assertEquals failed: want='$want' got='$got' $msg"
+    exit 2
+  fi
+}
+
+assertFilesDoesNotExist() {
+  path="$1"
+  msg=$2
+  if [ -f "${path}" ]; then
+    echo "assertFilesDoesNotExist failed: path exists '$path': $msg"
+    exit 2
+  fi
+}
+
+assertFileExists() {
+  path="$1"
+  msg=$2
+  if [ ! -f "${path}" ]; then
+    echo "assertFileExists failed: path does not exist '$path': $msg"
+    exit 2
+  fi
+}
+
+assertFilesEqual() {
+  want=$1
+  got=$2
+  msg=$3
+
+  diff "$1" "$2"
+  if [ $? -ne 0 ]; then
+    echo "assertFilesEqual failed: $msg"
+    exit 2
+  fi
+}
+
+assertNotEquals() {
+  want=$1
+  got=$2
+  msg=$3
+  if [ "$want" = "$got" ]; then
+    echo "assertNotEquals failed: want='$want' got='$got' $msg"
+    exit 2
+  fi
+}
+
+log_test_case() {
+  echo "  running $@"
+}
+
+run_test_case_with_snapshot_release() {
+  log_test_case ${@:1}
```

Comment:
In this case why does line 77 not execute, but line 83 will execute?

Reply:
in bash (not posix I think) when you pass items as parameters they are not evaluated, unless explicitly done so (e.g. `$($1)`). When used alone as not parameters they are implicitly evaluated --a more explicit approach would be to use `eval ${@:1}` I think.